lua/org-roam/database/loader.lua

```diff
@@ -373,46 +374,47 @@ end
 ---Loads database (or retrieves from cache) asynchronously.
 ---@return OrgPromise<org-roam.core.Database>
 function M:database()
-    return self.__db and Promise.resolve(self.__db)
-        or Promise.new(function(resolve)
-            -- Load our database from disk if it is available
-            io.stat(self.path.database)
-                :next(function()
-                    return Database:load_from_disk(self.path.database)
-                        :next(function(db)
-                            schema:update(db)
-                            self.__db = db
-
-                            resolve(db)
-                            return db
-                        end)
-                        :catch(function(err)
-                            log.fmt_error("Failed to load database: %s", err)
-
-                            -- Set up database with a clean slate instead
-                            local db = Database:new()
-                            schema:update(db)
-                            self.__db = db
-
-                            resolve(db)
-                        end)
-                end)
-                :catch(function()
-                    local db = Database:new()
-                    schema:update(db)
-                    self.__db = db
-
-                    return resolve(db)
-                end)
+    local state = self.__load_state
+    if state == nil then
+        local promise = io.stat(self.path.database)
+            :next(function()
+                return Database:load_from_disk(self.path.database)
+                    :next(function(db)
+                        schema:update(db)
+                        self.__db = db
+                        self.__load_state = true
+                        return nil
+                    end)
+                    :catch(function(err)
+                        log.fmt_error("Failed to load database: %s", err)
+                        -- Punt error to the outer `catch` so it can create the database.
+                        error(err)
+                    end)
+            end)
+            :catch(function()
+                local db = Database:new()
+                schema:update(db)
+                self.__db = db
+                self.__load_state = true
+            end)
+        self.__load_state = promise
```

Comment:
Why not set to promise:next with returning self.__db and then return self.__load_state?

You're having to repeat the next call on another line when state is not true or nil, anyway.